app/sample_manifest_excel/sample_manifest_excel/upload/processor/tube_rack.rb

```diff
@@ -0,0 +1,201 @@
+            return false if !existing_barcode_record.nil? && !existing_barcode_record.asset.nil?
+          end unless @tube_rack_barcodes_from_manifest.nil?
+          true
+        end
+
+        def retrieve_scan_results
+          @rack_barcode_to_scan_results = {}
+          @tube_barcode_to_rack_barcode = {}
+          @tube_rack_barcodes_from_manifest.each do |tube_rack_barcode|
+            results = if Rails.configuration.do_tube_rack_scan_callout
+                        retrieve_tube_rack_scan_from_microservice(tube_rack_barcode)
+                      else
+                        mock_scan_result
+                      end
+            return false if results.nil?
+
+            @rack_barcode_to_scan_results[tube_rack_barcode] = results
+            results.keys.each { |tube_barcode| @tube_barcode_to_rack_barcode[tube_barcode] = tube_rack_barcode }
+          end
+
+          true
+        end
+
+        def retrieve_tube_rack_scan_from_microservice(tube_rack_barcode)
+          # tube_rack_barcode = 'test_valid_file'
+          host_name = Rails.configuration.tube_rack_scans_microservice_endpoint
+          path = '/tube_rack/' + tube_rack_barcode
+          port = Rails.configuration.tube_rack_scans_microservice_port
+          response = Net::HTTP.get_response(host_name, path, port)
+
+          begin
+            scan_results = JSON.parse(response.body)
+          rescue JSON::JSONError => e
+            error_message = "Response when trying to retrieve scan (tube rack with barcode #{tube_rack_barcode}) was not valid JSON so could not be understood. Error message: #{e.message}"
+            upload.errors.add(:base, error_message)
+            return nil
+          end
+
+          unless response.code == '200'
+            error_message = "Scan could not be retrieved for tube rack with barcode #{tube_rack_barcode}. Service responded with status code #{response.code} "
+            error_message += " and the following message: #{scan_results['error']}"
+            upload.errors.add(:base, error_message)
+            return nil
+          end
+
+          scan_results['layout'] || nil
+        end
+
+        def mock_scan_result
+          {
+            'TB11111110' => 'e8',
+            'TB11111111' => 'b4'
+          }
+        end
+
+        def validate_against_scan_results
+          return true if @tube_barcodes_from_manifest.sort == @tube_barcode_to_rack_barcode.keys.sort
+
+          error_message = 'The scan and the manifest do not contain identical tube barcodes.'
+          upload.errors.add(:base, error_message)
+          false
+        end
+
+        def validate_coordinates(rack_size, rack_barcode_to_scan_results)
+          rack_layout_map = {
+            48 => {
+              'rows' => 6,
+              'columns' => 8
+            },
+            96 => {
+              'rows' => 8,
+              'columns' => 12
+            }
+          }
```

Comment:
I would move it somewhere as a constant, so it is easy to modify if we need to in future.